xds/internal/xdsclient/xdsresource/unmarshal_cds.go

```diff
@@ -463,3 +483,105 @@ func circuitBreakersFromCluster(cluster *v3clusterpb.Cluster) *uint32 {
 	}
 	return nil
 }
+
+// outlierConfigFromCluster extracts the relevant outlier detection configuration from
+// the received cluster resource. Returns nil if no OutlierDetection.
+func outlierConfigFromCluster(cluster *v3clusterpb.Cluster) (*OutlierDetection, error) {
+	// Corresponds to this line from A50: "If the outlier_detection field is not
+	// set in the Cluster message, a "no-op" outlier_detection config will be
+	// generated"
+	od := cluster.GetOutlierDetection()
+	if od == nil {
+		return nil, nil
+	}
+	// "The google.protobuf.Duration fields interval, base_ejection_time, and
+	// max_ejection_time must obey the restrictions in the
+	// google.protobuf.Duration documentation and they must have non-negative
+	// values." - A50
+	// TODO: I can find non negative examples for validation, but not "must obey the restrictions in the documentation"
+	var iVal, betVal, metVal = defaultInterval, defaultBaseEjectionTime, defaultMaxEjectionTime
+	if i := od.GetInterval(); i != nil {
+		iVal = i.AsDuration()
+		if iVal < 0 {
+			return nil, fmt.Errorf("outlier_detection.interval = %v; must be >= 0", iVal)
+		}
+	}
+
+	if bet := od.GetBaseEjectionTime(); bet != nil {
+		betVal = bet.AsDuration()
+		if betVal < 0 {
+			return nil, fmt.Errorf("outlier_detection.base_ejection_time = %v; must be >= 0", betVal)
+		}
+	}
+
+	if met := od.GetMaxEjectionTime(); met != nil {
+		metVal = met.AsDuration()
+		if metVal < 0 {
+			return nil, fmt.Errorf("outlier_detection.max_ejection_time = %v; must be >= 0", metVal)
+		}
+	}
+
+	var mepVal, esrVal, fptVal, efpVal, srsfVal, srmhVal, srrvVal, fpmhVal, fprvVal uint32 = defaultMaxEjectionPercent,
+		defaultEnforcingSuccessRate, defaultFailurePercentageThreshold, defaultEnforcingFailurePercentage, defaultSuccessRateStdevFactor,
+		defaultSuccessRateMinimumHosts, defaultSuccessRateRequestVolume, defaultFailurePercentageMinimumHosts,
+		defaultFailurePercentageRequestVolume
```

Comment:
Please define each variable exactly above the block where it is used. Same applies to the three vars defined earlier as well.